File: backend/internal/model/two_factor.go
package model

import "time"

// TOTPEnrollBegin is returned when a user starts 2FA enrollment.
// The QR code is rendered client-side from OtpauthURI; Secret is shown
// as a fallback for users who can't scan.
type TOTPEnrollBegin struct {
	Secret      string `json:"secret"`
	OtpauthURI  string `json:"otpauth_uri"`
	Issuer      string `json:"issuer"`
	AccountName string `json:"account_name"`
}

type TOTPEnrollConfirmRequest struct {
	Code string `json:"code"`
}

// TOTPDisableRequest accepts either the user's current password or a
// current TOTP/backup code to re-authenticate the disable action.
type TOTPDisableRequest struct {
	Password string `json:"password,omitempty"`
	Code     string `json:"code,omitempty"`
}

type TOTPVerifyLoginRequest struct {
	ChallengeToken string `json:"challenge_token"`
	Code           string `json:"code"`
}

type TOTPStatus struct {
	Enabled              bool       `json:"enabled"`
	EnrolledAt           *time.Time `json:"enrolled_at,omitempty"`
	BackupCodesRemaining int        `json:"backup_codes_remaining"`
}

type BackupCodesResponse struct {
	Codes []string `json:"codes"`
}

File: frontend/src/api/twoFactor.ts
import { api } from './client'
import type { User, AuthTokens } from './auth'

export interface TwoFactorStatus {
  enabled: boolean
  enrolled_at?: string
  backup_codes_remaining: number
}

export interface EnrollBeginResponse {
  secret: string
  otpauth_uri: string
  issuer: string
  account_name: string
}

export interface BackupCodesResponse {
  backup_codes: string[]
}

export interface CompleteLoginResponse {
  user: User
  tokens: AuthTokens
}

export const twoFactorApi = {
  getStatus: () =>
    api.get<TwoFactorStatus>('/users/me/2fa/status'),

  enrollBegin: () =>
    api.post<EnrollBeginResponse>('/users/me/2fa/enroll/begin', {}),

  enrollConfirm: (code: string) =>
    api.post<BackupCodesResponse>('/users/me/2fa/enroll/confirm', { code }),

  disable: (input: { password?: string; code?: string }) =>
    api.post<void>('/users/me/2fa/disable', input),

  regenerateBackupCodes: (code: string) =>
    api.post<BackupCodesResponse>('/users/me/2fa/backup-codes/regenerate', { code }),

  loginVerify2FA: (challengeToken: string, code: string) =>
    api.post<CompleteLoginResponse>('/auth/login/2fa', {
      challenge_token: challengeToken,
      code,
    }),
}
